<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.55">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Sequencing settings – tRNA-seq logistics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">tRNA-seq logistics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./background.html"> 
<span class="menu-text">Background</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./sequencing-settings.html" aria-current="page"> 
<span class="menu-text">Sequencing settings</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./library-structures.html"> 
<span class="menu-text">Library preparation &amp; structures</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#sequencing-settings-for-trna" id="toc-sequencing-settings-for-trna" class="nav-link active" data-scroll-target="#sequencing-settings-for-trna">Sequencing settings for tRNA</a>
  <ul class="collapse">
  <li><a href="#why-use-custom-sequencing-settings-for-trna" id="toc-why-use-custom-sequencing-settings-for-trna" class="nav-link" data-scroll-target="#why-use-custom-sequencing-settings-for-trna">Why use custom sequencing settings for tRNA?</a></li>
  <li><a href="#enabling-short-read-mode-in-minknow" id="toc-enabling-short-read-mode-in-minknow" class="nav-link" data-scroll-target="#enabling-short-read-mode-in-minknow">Enabling short read mode in MinKNOW</a>
  <ul class="collapse">
  <li><a href="#rna-short-mode-setup" id="toc-rna-short-mode-setup" class="nav-link" data-scroll-target="#rna-short-mode-setup">RNA short mode setup</a></li>
  <li><a href="#starting-a-trna-sequencing-run-in-short-mode" id="toc-starting-a-trna-sequencing-run-in-short-mode" class="nav-link" data-scroll-target="#starting-a-trna-sequencing-run-in-short-mode">Starting a tRNA sequencing run in short mode</a></li>
  <li><a href="#do-you-have-versions-of-these-files-i-can-just-use-instead-of-editing-code-by-hand" id="toc-do-you-have-versions-of-these-files-i-can-just-use-instead-of-editing-code-by-hand" class="nav-link" data-scroll-target="#do-you-have-versions-of-these-files-i-can-just-use-instead-of-editing-code-by-hand">Do you have versions of these files I can just use instead of editing code by hand?</a></li>
  </ul></li>
  <li><a href="#alternative-approach-simulating-runs-from-bulk-fast5-files" id="toc-alternative-approach-simulating-runs-from-bulk-fast5-files" class="nav-link" data-scroll-target="#alternative-approach-simulating-runs-from-bulk-fast5-files">Alternative approach: simulating runs from bulk fast5 files</a></li>
  <li><a href="#this-all-seems-complicated-cant-i-just-click-a-button-in-minknow-to-change-these-settings" id="toc-this-all-seems-complicated-cant-i-just-click-a-button-in-minknow-to-change-these-settings" class="nav-link" data-scroll-target="#this-all-seems-complicated-cant-i-just-click-a-button-in-minknow-to-change-these-settings">This all seems complicated, can’t I just click a button in MinKNOW to change these settings?</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Sequencing settings</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="sequencing-settings-for-trna" class="level1">
<h1>Sequencing settings for tRNA</h1>
<p>MinKNOW is the operating software that coordinates setting up a sequencing run on an ONT sequencing device (e.g., a MinION or PromethION). It captures and optionally basecalls your data during the run, and monitors and responds to feedback from flow cells in real time. The first step in setting up a run is selecting the appropriate settings (e.g., making sure the parameters match the flow cell type and sequencing kit you used to prepare and load your library). For tRNA, this involves some extra steps.</p>
<section id="why-use-custom-sequencing-settings-for-trna" class="level2">
<h2 class="anchored" data-anchor-id="why-use-custom-sequencing-settings-for-trna">Why use custom sequencing settings for tRNA?</h2>
<p>Back when we first started making tRNA sequencing libraries using the approach from Thomas <em>et al.</em> 2021 and running them with default parameters, our runs looked like this:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/2022_tRNA_samplerun.png" class="img-fluid figure-img"></p>
<figcaption>“tRNA sample run”</figcaption>
</figure>
</div>
<p>Note the high proportion of “adapter-only” reads (beige bar) relative to the bright green bar that MinKNOW believes are genuine sequencing reads. <a href="https://pubmed.ncbi.nlm.nih.gov/37024678/">Lucas <em>et. al</em> 2023</a> expertly dissects the reasons behind this issue. In brief, MinKNOW’s default settings expect the <em>adapter</em> portion of reads (or adapter-only reads) to transit through the pore in 5 seconds or less, and expect the remaining portion of the RNA molecule (that is, the thing you’re actually looking to sequence), termed the <em>strand</em>, to take at least 2 seconds to pass through the pore. These default settings from ONT were selected with mRNA sequencing in mind.</p>
<p>tRNA molecules are much shorter than mRNA and in many cases, the strand portion of the read can pass through the pore in under two seconds. The Novoa lab demonstrated that lowering the minimum threshold at which MinKNOW classifies the <em>strand</em> portion of read from 2 seconds to 1 second, combined with lowering the maximum duration for <em>adapter</em> signal from 5 seconds to 2 second, increased the throughput of tRNA sequencing by over an order of magnitude, suggesting that the default RNA sequencing parameters classify many tRNA reads as free adapter.</p>
<p><strong>TLDR: If you don’t want most of your reads to be inadvertedly discarded, you need to take some extra steps </strong>before** you begin sequencing tRNA libraries.**</p>
<p>NB: To the best of my knowledge, MinKNOW senses the break between <em>adapter</em> and <em>strand</em> portions of a read by looking for signal from the RTA adapter; by default, neither this adapter sequence nor the adjoining poly(A) tail will be incorporated into the FASTQ produced during basecalling. In contrast, the sequence from the double stranded adapters ligated to tRNA ends in the first step of nanopore sequencing protocols <em>is</em> basecalled and incorporated into the read sequence stored in the FASTQ files.</p>
</section>
<section id="enabling-short-read-mode-in-minknow" class="level2">
<h2 class="anchored" data-anchor-id="enabling-short-read-mode-in-minknow">Enabling short read mode in MinKNOW</h2>
<section id="rna-short-mode-setup" class="level3">
<h3 class="anchored" data-anchor-id="rna-short-mode-setup">RNA short mode setup</h3>
<p>We wanted to be able to use a custom MinKNOW configuration for tRNA in real time during sequencing. To do this, we need to make a few changes to the MinKNOW software.</p>
<p>MinKNOW’s configuration files live at the following locations:</p>
<ul>
<li>In Linux: <code>/opt/ont/minknow/conf/package/</code></li>
<li>In MacOS: <code>/Applications/MinKNOW.app/Contents/Resources/conf/package/</code></li>
</ul>
<p>You will need root permissions to edit and add files in the <code>/package/</code> folder. We have done this as follows:</p>
<ul>
<li>At the bottom of <code>flow_cells.toml</code>, add the following lines:</li>
</ul>
<pre><code># custom configs

[FLO-FLG001_short]
connector = "flongle"

[FLO-MIN106_short]
connector = "minion_mk1"

[FLO-PRO002_short]
connector = "promethion"</code></pre>
<ul>
<li>Navigate to the <code>/sequencing/</code> folder and identify the <code>.toml</code> files for direct RNA sequencing (<code>sequencing_MIN106_RNA.toml</code> for MinION and <code>sequencing_PRO002_RNA.toml</code> for PromethION). Copy the relevant file(s) to create a second version appended <code>_short</code>, e.g.:</li>
</ul>
<pre><code>sudo cp sequencing_PRO002_RNA.toml sequencing_PRO002_RNA_short.toml </code></pre>
<ul>
<li>Open the new <code>_short</code> version of the <code>RNA.toml</code> file and find the section containing the Read Classification Parameters. It should look like this:</li>
</ul>
<pre><code>[analysis_configuration.read_classification.parameters]
rules_in_execution_order = [
     "multiple=        (median,gt,350)&amp;(median,lt,990)&amp;(local_median_sd,gt,2)&amp;(duration,gt,0.1)",
     "pore_1=          (local_median,gt,160)&amp;(local_median,lt,280)&amp;(median_sd,gt,0.9)&amp;(median_sd,lt,5)&amp;(local_range,lt,35)&amp;(duration,gt,15)",
     "pore=            (median,gt,160)&amp;(median,lt,280)&amp;(median_sd,gt,0.6)&amp;(median_sd,lt,5)",
     "event=           (median_before,gt,160)&amp;(median_before,lt,280)&amp;(median_before,gt,median)&amp;(median,lt,160)&amp;(median,gt,20)&amp;(duration,lt,0.1)&amp;(event_count,lt,10)",
     "adapter=         (median_before,gt,160)&amp;(median_before,lt,280)&amp;(local_range,gt,5)&amp;(local_range,lt,50)&amp;(local_median,gt,50)&amp;(local_median,lt,120)&amp;(local_median_sd,gt,0.5)&amp;(local_median_sd,lt,2.5)&amp;(duration,lt,5)",
     "strand=          (local_range,gt,25)&amp;(local_range,lt,60)&amp;(local_median,gt,60)&amp;(local_median,lt,115)&amp;(local_median_sd,gt,1)&amp;(local_median_sd,lt,4)&amp;(duration,gt,2)",</code></pre>
<ul>
<li>Edit the last number on each of the final lines above so that the duration for <code>adapter</code> ends in <code>lt,2</code> and the duration for <code>strand</code> ends in <code>gt,1</code>. These are the thresholds in seconds described above.</li>
</ul>
<p>The approach above has been tested for R9.4 Flongle and MinION flow cells running on a MK1B MinION connected to M1 Macbook Pro running MacOS Ventura, and on R9.4 flow cells running on a beta PromethION 24 connected to a desktop running Ubuntu 20.04. Both of these combinations were tested with MinKNOW version 23.04.3.</p>
</section>
<section id="starting-a-trna-sequencing-run-in-short-mode" class="level3">
<h3 class="anchored" data-anchor-id="starting-a-trna-sequencing-run-in-short-mode">Starting a tRNA sequencing run in short mode</h3>
<p>After making the changes above and restarting MinKNOW, <code>FLO-MIN106_short</code> and <code>FLO-PRO002_short</code> will be available as drop down options under Flow Cell Type. Once you select one of these, Direct RNA Sequencing should be the only kit available on the subsequent page. You can set up the rest of the sequencing run in MinKNOW as normal.</p>
</section>
<section id="do-you-have-versions-of-these-files-i-can-just-use-instead-of-editing-code-by-hand" class="level3">
<h3 class="anchored" data-anchor-id="do-you-have-versions-of-these-files-i-can-just-use-instead-of-editing-code-by-hand">Do you have versions of these files I can just use instead of editing code by hand?</h3>
<p><strong>Yes.</strong> You can find them in the companion repository <a href="https://github.com/lkwhite/tRNAseq/conf">here</a>, but caveat emptor. The MinKNOW software undergoes <a href="https://community.nanoporetech.com/downloads">frequent updates</a>, and using an outdated <code>.toml</code> file may break your ability to run RNA short mode live. At a minimum, you will want to update the <code>[compatibility]</code> section of the <code>.toml</code> to match the MinKNOW core and Bream versions to whatever version of MinKNOW you’re using. You can find these in the MinKNOW UI under Host Settings &gt; Software, and then mousing over the text Installed Version to get a text box on hover.</p>
</section>
</section>
<section id="alternative-approach-simulating-runs-from-bulk-fast5-files" class="level2">
<h2 class="anchored" data-anchor-id="alternative-approach-simulating-runs-from-bulk-fast5-files">Alternative approach: simulating runs from bulk fast5 files</h2>
<p>The Novoa lab’s <a href="https://pubmed.ncbi.nlm.nih.gov/37024678/">Lucas <em>et. al</em> 2023</a> has an accompanying <a href="https://github.com/novoalab/Nano-tRNAseq/tree/main/conf">README within the nano-tRNAseq repository</a> for setting up alternative configurations.</p>
<p>As present, this workflow involves:</p>
<ul>
<li>rsyncing custom <code>*RNA_short.toml</code> and <code>flow_cells.toml</code> config files containing the same changes to duration parameters described above</li>
<li>saving out a bulk fast5 file</li>
<li>editing the <code>*RNA_short.toml</code> file to add a path to the bulk fast5 file</li>
<li>running a sequencing simulation in MinKNOW from the bulk fast5 file</li>
</ul>
<p>The .toml config files in this repo are from a much older version of MinKNOW, and we couldn’t get them working for real time tRNA sequencing. Since sequencing simulations are pretty cumbersome and bulk fast5 files take up a lot of space, we opted instead to edit more recent versions of the sequencing <code>.toml</code> files and run short mode live.</p>
<p>However, if you’re interested in running simulations to test different MinKNOW parameters on the same data, I recommend checking the nano-tRNAseq repository above as well as the documentation for <a href="https://github.com/LooseLab/readfish">ReadFish</a> from Matt Loose’s lab. <strong>Pay particular attention to the list of GOTCHAs on the ReadFish README!</strong> If you forget you’re running a <code>.toml</code> file that contains a simulation path to a bulk fast5, bad things can and likely will happen.</p>
</section>
<section id="this-all-seems-complicated-cant-i-just-click-a-button-in-minknow-to-change-these-settings" class="level2">
<h2 class="anchored" data-anchor-id="this-all-seems-complicated-cant-i-just-click-a-button-in-minknow-to-change-these-settings">This all seems complicated, can’t I just click a button in MinKNOW to change these settings?</h2>
<p><strong>Not yet.</strong></p>
<p>A short fragment mode was released as a selectable option for nanopore <a href="https://nanoporetech.com/applications/techniques/short-fragment-mode">DNA sequencing in March 2022</a>; we expect a similar mode to eventually be made available for direct RNA sequencing runs in a future version of the MinKNOW software.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/rnabioco\.github\.io\/tRNAseq-logistics\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>