<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <link rel="stylesheet" href="/tRNAseq/assets/css/just-the-docs-default.css"> <link rel="stylesheet" href="/tRNAseq/assets/css/just-the-docs-head-nav.css" id="jtd-head-nav-stylesheet"> <style id="jtd-nav-activation"> .site-nav > ul.nav-list:first-child > li:not(:nth-child(4)) > a, .site-nav > ul.nav-list:first-child > li > ul > li > a, .site-nav > ul.nav-list:first-child > li > ul > li > ul > li > a { background-image: none; } .site-nav > ul.nav-list:not(:first-child) a, .site-nav li.external a { background-image: none; } .site-nav > ul.nav-list:first-child > li:nth-child(4) > a { font-weight: 600; text-decoration: none; }.site-nav > ul.nav-list:first-child > li:nth-child(4) > button svg { transform: rotate(-90deg); }.site-nav > ul.nav-list:first-child > li.nav-list-item:nth-child(4) > ul.nav-list { display: block; } </style> <script src="/tRNAseq/assets/js/vendor/lunr.min.js"></script> <script src="/tRNAseq/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.8.0 --> <title>Sequencing settings | A tRNAseq guide</title> <meta name="generator" content="Jekyll v4.3.3" /> <meta property="og:title" content="Sequencing settings" /> <meta property="og:locale" content="en_US" /> <meta name="description" content="Notes on library prep and analysis" /> <meta property="og:description" content="Notes on library prep and analysis" /> <link rel="canonical" href="https://lkwhite.github.io/tRNAseq/sequencing-settings/" /> <meta property="og:url" content="https://lkwhite.github.io/tRNAseq/sequencing-settings/" /> <meta property="og:site_name" content="A tRNAseq guide" /> <meta property="og:type" content="website" /> <meta name="twitter:card" content="summary" /> <meta property="twitter:title" content="Sequencing settings" /> <script type="application/ld+json"> {"@context":"https://schema.org","@type":"WebPage","description":"Notes on library prep and analysis","headline":"Sequencing settings","url":"https://lkwhite.github.io/tRNAseq/sequencing-settings/"}</script> <!-- End Jekyll SEO tag --> </head> <body> <a class="skip-to-main" href="#main-content">Skip to main content</a> <svg xmlns="http://www.w3.org/2000/svg" class="d-none"> <symbol id="svg-link" viewBox="0 0 24 24"> <title>Link</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"> <path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"> <title>Menu</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"> <line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"> <title>Expand</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"> <polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <!-- Feather. MIT License: https://github.com/feathericons/feather/blob/master/LICENSE --> <symbol id="svg-external-link" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-external-link"> <title id="svg-external-link-title">(external link)</title> <path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"></path><polyline points="15 3 21 3 21 9"></polyline><line x1="10" y1="14" x2="21" y2="3"></line> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"> <title>Document</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"> <path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"> <title>Search</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <!-- Bootstrap Icons. MIT License: https://github.com/twbs/icons/blob/main/LICENSE.md --> <symbol id="svg-copy" viewBox="0 0 16 16"> <title>Copy</title> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard" viewBox="0 0 16 16"> <path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1h1a1 1 0 0 1 1 1V14a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V3.5a1 1 0 0 1 1-1h1v-1z"/> <path d="M9.5 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3zm-3-1A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3z"/> </svg> </symbol> <symbol id="svg-copied" viewBox="0 0 16 16"> <title>Copied</title> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clipboard-check-fill" viewBox="0 0 16 16"> <path d="M6.5 0A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3Zm3 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3Z"/> <path d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1A2.5 2.5 0 0 1 9.5 5h-3A2.5 2.5 0 0 1 4 2.5v-1Zm6.854 7.354-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 0 1 .708-.708L7.5 10.793l2.646-2.647a.5.5 0 0 1 .708.708Z"/> </svg> </symbol> </svg> <div class="side-bar"> <div class="site-header" role="banner"> <a href="/tRNAseq/" class="site-title lh-tight"> A tRNAseq guide </a> <button id="menu-button" class="site-button btn-reset" aria-label="Toggle menu" aria-pressed="false"> <svg viewBox="0 0 24 24" class="icon" aria-hidden="true"><use xlink:href="#svg-menu"></use></svg> </button> </div> <nav aria-label="Main" id="site-nav" class="site-nav"> <ul class="nav-list"><li class="nav-list-item"><a href="/tRNAseq/" class="nav-list-link">Home</a></li><li class="nav-list-item"><a href="/tRNAseq/background/" class="nav-list-link">Background</a></li><li class="nav-list-item"><a href="/tRNAseq/library-structures/" class="nav-list-link">Library preparation & structures</a></li><li class="nav-list-item"><a href="/tRNAseq/sequencing-settings/" class="nav-list-link">Sequencing settings</a></li></ul> </nav> <footer class="site-footer"> This site uses <a href="https://github.com/just-the-docs/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll. </footer> </div> <div class="main" id="top"> <div id="main-header" class="main-header"> <div class="search" role="search"> <div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search A tRNAseq guide" aria-label="Search A tRNAseq guide" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label> </div> <div id="search-results" class="search-results"></div> </div> <nav aria-label="Auxiliary" class="aux-nav"> <ul class="aux-nav-list"> <li class="aux-nav-list-item"> <a href="https://github.com/lkwhite/tRNAseq" class="site-button" > GH Repository </a> </li> </ul> </nav> </div> <div class="main-content-wrap"> <div id="main-content" class="main-content"> <main> <h1 id="sequencing-settings-for-trna"> <a href="#sequencing-settings-for-trna" class="anchor-heading" aria-labelledby="sequencing-settings-for-trna"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Sequencing settings for tRNA </h1> <p>MinKNOW is the operating software that coordinates setting up a sequencing run on an ONT sequencing device (e.g., a MinION or PromethION). It captures and optionally basecalls your data during the run, and monitors and responds to feedback from flow cells in real time. The first step in setting up a run is selecting the appropriate settings (e.g., making sure the parameters match the flow cell type and sequencing kit you used to prepare and load your library). For tRNA, this involves some extra steps.</p> <h2 id="why-use-custom-sequencing-settings-for-trna"> <a href="#why-use-custom-sequencing-settings-for-trna" class="anchor-heading" aria-labelledby="why-use-custom-sequencing-settings-for-trna"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Why use custom sequencing settings for tRNA? </h2> <p>Back when we first started making tRNA sequencing libraries using the approach from Thomas <em>et al.</em> 2021 and running them with default parameters, our runs looked like this:</p> <p><img src="https://lkwhite.github.io/tRNAseq/docs/assets/images/2022_tRNA_samplerun.png" alt="tRNA Sample Run" /></p> <p>Note the high proportion of “adapter-only” reads (beige bar) relative to the bright green bar that MinKNOW believes are genuine sequencing reads. <a href="https://pubmed.ncbi.nlm.nih.gov/37024678/">Lucas <em>et. al</em> 2023</a> expertly dissects the reasons behind this issue. In brief, MinKNOW’s default settings expect the <em>adapter</em> portion of reads (or adapter-only reads) to transit through the pore in 5 seconds or less, and expect the remaining portion of the RNA molecule (that is, the thing you’re actually looking to sequence), termed the <em>strand</em>, to take at least 2 seconds to pass through the pore. These default settings from ONT were selected with mRNA sequencing in mind.</p> <p>tRNA molecules are much shorter than mRNA and in many cases, the strand portion of the read can pass through the pore in under two seconds. The Novoa lab demonstrated that lowering the minimum threshold at which MinKNOW classifies the <em>strand</em> portion of read from 2 seconds to 1 second, combined with lowering the maximum duration for <em>adapter</em> signal from 5 seconds to 2 second, increased the throughput of tRNA sequencing by over an order of magnitude, suggesting that the default RNA sequencing parameters classify many tRNA reads as free adapter.</p> <p><strong>TLDR: If you don’t want most of your reads to be inadvertedly discarded, you need to take some extra steps **before</strong> you begin sequencing tRNA libraries.**</p> <p>NB: To the best of my knowledge, MinKNOW senses the break between <em>adapter</em> and <em>strand</em> portions of a read by looking for signal from the RTA adapter; by default, neither this adapter sequence nor the adjoining poly(A) tail will be incorporated into the FASTQ produced during basecalling. In contrast, the sequence from the double stranded adapters ligated to tRNA ends in the first step of nanopore sequencing protocols <em>is</em> basecalled and incorporated into the read sequence stored in the FASTQ files.</p> <h2 id="enabling-short-read-mode-in-minknow"> <a href="#enabling-short-read-mode-in-minknow" class="anchor-heading" aria-labelledby="enabling-short-read-mode-in-minknow"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Enabling short read mode in MinKNOW </h2> <h3 id="rna-short-mode-setup"> <a href="#rna-short-mode-setup" class="anchor-heading" aria-labelledby="rna-short-mode-setup"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> RNA short mode setup </h3> <p>We wanted to be able to use a custom MinKNOW configuration for tRNA in real time during sequencing. To do this, we need to make a few changes to the MinKNOW software.</p> <p>MinKNOW’s configuration files live at the following locations:</p> <ul> <li>In Linux: <code class="language-plaintext highlighter-rouge">/opt/ont/minknow/conf/package/</code></li> <li>In MacOS: <code class="language-plaintext highlighter-rouge">/Applications/MinKNOW.app/Contents/Resources/conf/package/</code></li> </ul> <p>You will need root permissions to edit and add files in the <code class="language-plaintext highlighter-rouge">/package/</code> folder. We have done this as follows:</p> <ul> <li>At the bottom of <code class="language-plaintext highlighter-rouge">flow_cells.toml</code>, add the following lines:</li> </ul> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># custom configs

[FLO-FLG001_short]
connector = "flongle"

[FLO-MIN106_short]
connector = "minion_mk1"

[FLO-PRO002_short]
connector = "promethion"
</code></pre></div></div> <ul> <li>Navigate to the <code class="language-plaintext highlighter-rouge">/sequencing/</code> folder and identify the <code class="language-plaintext highlighter-rouge">.toml</code> files for direct RNA sequencing (<code class="language-plaintext highlighter-rouge">sequencing_MIN106_RNA.toml</code> for MinION and <code class="language-plaintext highlighter-rouge">sequencing_PRO002_RNA.toml</code> for PromethION). Copy the relevant file(s) to create a second version appended <code class="language-plaintext highlighter-rouge">_short</code>, e.g.:</li> </ul> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sudo cp sequencing_PRO002_RNA.toml sequencing_PRO002_RNA_short.toml 
</code></pre></div></div> <ul> <li>Open the new <code class="language-plaintext highlighter-rouge">_short</code> version of the <code class="language-plaintext highlighter-rouge">RNA.toml</code> file and find the section containing the Read Classification Parameters. It should look like this:</li> </ul> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[analysis_configuration.read_classification.parameters]
rules_in_execution_order = [
     "multiple=        (median,gt,350)&amp;(median,lt,990)&amp;(local_median_sd,gt,2)&amp;(duration,gt,0.1)",
     "pore_1=          (local_median,gt,160)&amp;(local_median,lt,280)&amp;(median_sd,gt,0.9)&amp;(median_sd,lt,5)&amp;(local_range,lt,35)&amp;(duration,gt,15)",
     "pore=            (median,gt,160)&amp;(median,lt,280)&amp;(median_sd,gt,0.6)&amp;(median_sd,lt,5)",
     "event=           (median_before,gt,160)&amp;(median_before,lt,280)&amp;(median_before,gt,median)&amp;(median,lt,160)&amp;(median,gt,20)&amp;(duration,lt,0.1)&amp;(event_count,lt,10)",
     "adapter=         (median_before,gt,160)&amp;(median_before,lt,280)&amp;(local_range,gt,5)&amp;(local_range,lt,50)&amp;(local_median,gt,50)&amp;(local_median,lt,120)&amp;(local_median_sd,gt,0.5)&amp;(local_median_sd,lt,2.5)&amp;(duration,lt,5)",
     "strand=          (local_range,gt,25)&amp;(local_range,lt,60)&amp;(local_median,gt,60)&amp;(local_median,lt,115)&amp;(local_median_sd,gt,1)&amp;(local_median_sd,lt,4)&amp;(duration,gt,2)",
</code></pre></div></div> <ul> <li>Edit the last number on each of the final lines above so that the duration for <code class="language-plaintext highlighter-rouge">adapter</code> ends in <code class="language-plaintext highlighter-rouge">lt,2</code> and the duration for <code class="language-plaintext highlighter-rouge">strand</code> ends in <code class="language-plaintext highlighter-rouge">gt,1</code>. These are the thresholds in seconds described above.</li> </ul> <p>The approach above has been tested for R9.4 Flongle and MinION flow cells running on a MK1B MinION connected to M1 Macbook Pro running MacOS Ventura, and on R9.4 flow cells running on a beta PromethION 24 connected to a desktop running Ubuntu 20.04. Both of these combinations were tested with MinKNOW version 23.04.3.</p> <h3 id="starting-a-trna-sequencing-run-in-short-mode"> <a href="#starting-a-trna-sequencing-run-in-short-mode" class="anchor-heading" aria-labelledby="starting-a-trna-sequencing-run-in-short-mode"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Starting a tRNA sequencing run in short mode </h3> <p>After making the changes above and restarting MinKNOW, <code class="language-plaintext highlighter-rouge">FLO-MIN106_short</code> and <code class="language-plaintext highlighter-rouge">FLO-PRO002_short</code> will be available as drop down options under Flow Cell Type. Once you select one of these, Direct RNA Sequencing should be the only kit available on the subsequent page. You can set up the rest of the sequencing run in MinKNOW as normal.</p> <h3 id="do-you-have-versions-of-these-files-i-can-just-use-instead-of-editing-code-by-hand"> <a href="#do-you-have-versions-of-these-files-i-can-just-use-instead-of-editing-code-by-hand" class="anchor-heading" aria-labelledby="do-you-have-versions-of-these-files-i-can-just-use-instead-of-editing-code-by-hand"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Do you have versions of these files I can just use instead of editing code by hand? </h3> <p><strong>Yes.</strong> You can find them in the companion repository <a href="https://github.com/lkwhite/tRNAseq/conf">here</a>, but caveat emptor. The MinKNOW software undergoes <a href="https://community.nanoporetech.com/downloads">frequent updates</a>, and using an outdated <code class="language-plaintext highlighter-rouge">.toml</code> file may break your ability to run RNA short mode live. At a minimum, you will want to update the <code class="language-plaintext highlighter-rouge">[compatibility]</code> section of the <code class="language-plaintext highlighter-rouge">.toml</code> to match the MinKNOW core and Bream versions to whatever version of MinKNOW you’re using. You can find these in the MinKNOW UI under Host Settings &gt; Software, and then mousing over the text Installed Version to get a text box on hover.</p> <h2 id="alternative-approach-simulating-runs-from-bulk-fast5-files"> <a href="#alternative-approach-simulating-runs-from-bulk-fast5-files" class="anchor-heading" aria-labelledby="alternative-approach-simulating-runs-from-bulk-fast5-files"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Alternative approach: simulating runs from bulk fast5 files </h2> <p>The Novoa lab’s <a href="https://pubmed.ncbi.nlm.nih.gov/37024678/">Lucas <em>et. al</em> 2023</a> has an accompanying <a href="https://github.com/novoalab/Nano-tRNAseq/tree/main/conf">README within the nano-tRNAseq repository</a> for setting up alternative configurations.</p> <p>As present, this workflow involves:</p> <ul> <li>rsyncing custom <code class="language-plaintext highlighter-rouge">*RNA_short.toml</code> and <code class="language-plaintext highlighter-rouge">flow_cells.toml</code> config files containing the same changes to duration parameters described above</li> <li>saving out a bulk fast5 file</li> <li>editing the <code class="language-plaintext highlighter-rouge">*RNA_short.toml</code> file to add a path to the bulk fast5 file</li> <li>running a sequencing simulation in MinKNOW from the bulk fast5 file</li> </ul> <p>The .toml config files in this repo are from a much older version of MinKNOW, and we couldn’t get them working for real time tRNA sequencing. Since sequencing simulations are pretty cumbersome and bulk fast5 files take up a lot of space, we opted instead to edit more recent versions of the sequencing <code class="language-plaintext highlighter-rouge">.toml</code> files and run short mode live.</p> <p>However, if you’re interested in running simulations to test different MinKNOW parameters on the same data, I recommend checking the nano-tRNAseq repository above as well as the documentation for <a href="https://github.com/LooseLab/readfish">ReadFish</a> from Matt Loose’s lab. <strong>Pay particular attention to the list of GOTCHAs on the ReadFish README!</strong> If you forget you’re running a <code class="language-plaintext highlighter-rouge">.toml</code> file that contains a simulation path to a bulk fast5, bad things can and likely will happen.</p> <h2 id="this-all-seems-complicated-cant-i-just-click-a-button-in-minknow-to-change-these-settings"> <a href="#this-all-seems-complicated-cant-i-just-click-a-button-in-minknow-to-change-these-settings" class="anchor-heading" aria-labelledby="this-all-seems-complicated-cant-i-just-click-a-button-in-minknow-to-change-these-settings"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> This all seems complicated, can’t I just click a button in MinKNOW to change these settings? </h2> <p><strong>Not yet.</strong></p> <p>A short fragment mode was released as a selectable option for nanopore <a href="https://nanoporetech.com/applications/techniques/short-fragment-mode">DNA sequencing in March 2022</a>; we expect a similar mode to eventually be made available for direct RNA sequencing runs in a future version of the MinKNOW software.</p> </main> </div> </div> <div class="search-overlay"></div> </div> </body> </html>
